File: bindings/nodejs/lib/tuple.js
/*
 * tuple.js
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


"use strict";

var assert = require('assert');
var buffer = require('./bufferConversion');
var fdbUtil = require('./fdbUtil');
var fdb = require('./fdbModule');
var FDBError = require('./error');

var sizeLimits = new Array(8);

function setupSizeLimits() {
	sizeLimits[0] = 1;
	for(var i = 1; i < sizeLimits.length; i++) {
		sizeLimits[i] = sizeLimits[i-1] * 256;
		sizeLimits[i-1] -= 1;
	}
	sizeLimits[7] -= 1;
}

setupSizeLimits();

var maxInt = Math.pow(2, 53) - 1;
var minInt = -Math.pow(2, 53);

var nullByte = new Buffer('00', 'hex');

var BYTES_CODE    = 0x01;
var STRING_CODE   = 0x02;
var NESTED_CODE   = 0x05;
var INT_ZERO_CODE = 0x14;
var POS_INT_END   = 0x1d;
var NEG_INT_END   = 0x0b;
var FLOAT_CODE    = 0x20;
var DOUBLE_CODE   = 0x21;
var FALSE_CODE    = 0x26;
var TRUE_CODE     = 0x27;
var UUID_CODE     = 0x30;

function validateData(data, length) {
	if(!(data instanceof Buffer) && !(data instanceof Array)) {
		throw new TypeError('Data for FDB tuple type not array or buffer.');
	} else if(data.length != length) {
		throw new RangeError('Data for FDB tuple type has length ' + data.length + ' instead of expected length ' + length);
	}
}

// If encoding and sign bit is 1 (negative), flip all of the bits. Otherwise, just flip sign.
// If decoding and sign bit is 0 (negative), flip all of the bits. Otherwise, just flip sign.
function adjustFloat(data, start, encode) {
	if((encode && (data[start] & 0x80) === 0x80) || (!encode && (data[start] & 0x80) === 0x00)) {
		for(var i = start; i < data.length; i++) {
			data[i] = data[i] ^ 0xff;
		}
	} else {
		data[start] = data[start] ^ 0x80;
	}
}

function Float(value) {
	this.value = value;
	this.toBytes = function () {
		if (this.rawData !== undefined) {
			return this.rawData;
		} else {
			var buf = new Buffer(4);
			buf.writeFloatBE(fdb.toFloat(this.value), 0);
			return buf;
		}
	};
}

Float.fromBytes = function (buf) {
	validateData(buf, 4);
	var f = new Float(buf.readFloatBE(0));
	if(isNaN(f.value)) {
		f.rawData = buf;
	}
	return f;
}

function Double(value) {
	this.value = value;
	this.toBytes = function () {
		if (this.rawData !== undefined) {
			return this.rawData;
		} else {
			var buf = new Buffer(8);
			buf.writeDoubleBE(this.value, 0);
			return buf;
		}
	};
}

Double.fromBytes = function (buf) {
	validateData(buf, 8);
	var d = new Double(buf.readDoubleBE(0));
	if(isNaN(d.value)) {
		d.rawData = buf;
	}
	return d;
}

function UUID(data) {
    if (data.length != 16) {
        // There's a special code for this, so we check it first and throw the error if appropriate.
        throw new FDBError("invalid_uuid_size", 2268);
    }
	validateData(data, 16);
	this.data = new Buffer(data);
}

function findNullBytes(buf, pos, searchForTerminators) {
	var nullBytes = [];

	var found;
	for(pos; pos < buf.length; ++pos) {
		if(searchForTerminators && found && buf[pos] !== 255) {
			break;
		}

		found = false;
		if(buf[pos] === 0) {
			found = true;
			nullBytes.push(pos);
		}
	}

	if(!found && searchForTerminators) {
		nullBytes.push(buf.length);
	}

	return nullBytes;
}

function encode(item, buf, pos) {
	var encodedString;
	if(typeof item === 'undefined')
		throw new TypeError('Packed element cannot be undefined');

	else if(item === null)
		return nullByte;

	//byte string or unicode
	else if(Buffer.isBuffer(item) || item instanceof ArrayBuffer || item instanceof Uint8Array || typeof item === 'string') {
		var unicode = typeof item === 'string';

		if(unicode) {
			item = new Buffer(item, 'utf8');
		}
		else {
			item = buffer(item);
		}

		var nullBytes = findNullBytes(item, 0);

		encodedString = new Buffer(2 + item.length + nullBytes.length);
		encodedString[0] = unicode ? STRING_CODE : BYTES_CODE;

		var srcPos = 0;
		var targetPos = 1;
		for(var i = 0; i < nullBytes.length; ++i) {
			item.copy(encodedString, targetPos, srcPos, nullBytes[i]+1);
			targetPos += nullBytes[i]+1 - srcPos;
			srcPos = nullBytes[i]+1;
			encodedString[targetPos++] = 255;
		}

		item.copy(encodedString, targetPos, srcPos);
		encodedString[encodedString.length-1] = 0;

		return encodedString;
	}

	//64-bit integer
	else if((typeof item === 'number' || item instanceof Number) && item % 1 === 0) {
		var negative = item < 0;
		var posItem = Math.abs(item);

		var length = 0;
		for(; length < sizeLimits.length; ++length) {
			if(posItem <= sizeLimits[length])
				break;
		}

		if(item > maxInt || item < minInt)
			throw new RangeError('Cannot pack signed integer larger than 54 bits');

		var prefix = negative ? INT_ZERO_CODE - length : INT_ZERO_CODE + length;

		var outBuf = new Buffer(length+1);
		outBuf[0] = prefix;
		for(var byteIdx = length-1; byteIdx >= 0; --byteIdx) {
			var b = posItem & 0xff;
			if(negative)
				outBuf[byteIdx+1] = ~b;
			else {
				outBuf[byteIdx+1] = b;
			}

			posItem = (posItem - b) / 0x100;
		}

		return outBuf;
	}

	// Floats
	else if(item instanceof Float) {
		var outBuf = new Buffer(5);
		outBuf[0] = FLOAT_CODE;
		if (isNaN(item.value) && item.rawData !== undefined) {
			item.rawData.copy(outBuf, 1, 0, 4);
		} else {
			outBuf.writeFloatBE(fdb.toFloat(item.value), 1);
		}
		adjustFloat(outBuf, 1, true);
		return outBuf;
	}

	// Doubles
	else if(item instanceof Double) {
		var outBuf = new Buffer(9);
		outBuf[0] = DOUBLE_CODE;
		if (isNaN(item.value) && item.rawData !== undefined) {
			item.rawData.copy(outBuf, 1, 0, 8);
		} else {
			outBuf.writeDoubleBE(item.value, 1);
		}
		adjustFloat(outBuf, 1, true);
		return outBuf;
	}

	// UUIDs
	else if(item instanceof UUID) {
		var outBuf = new Buffer(17);
		outBuf[0] = UUID_CODE;
		item.data.copy(outBuf, 1);
		return outBuf;
	}

	// booleans
	else if(item instanceof Boolean || typeof item === 'boolean') {
		var outBuf = new Buffer(1);
		var boolItem;
		if(item instanceof Boolean) {
			boolItem = item.valueOf();
		} else {
			boolItem = item;
		}
		if(boolItem) {
			outBuf[0] = TRUE_CODE;
		} else {
			outBuf[0] = FALSE_CODE;
		}
		return outBuf;
	}

	// nested tuples
	else if(item instanceof Array) {
		var totalLength = 2;
		var outArr = [new Buffer('05', 'hex')];
		for(var i = 0; i < item.length; ++i) {
			if(item[i] === null) {
				outArr.push(new Buffer('00ff', 'hex'));
				totalLength += 2;
			} else {
				outArr.push(encode(item[i]));
				totalLength += outArr[i+1].length;
			}
		}
		outArr.push(new Buffer('00', 'hex'))
		return Buffer.concat(outArr, totalLength);
	}

	else
		throw new TypeError('Packed element must either be a string, a buffer, an integer, or null');
}

function pack(arr) {
	if(!(arr instanceof Array))
		throw new TypeError('fdb.tuple.pack must be called with a single array argument');

	var totalLength = 0;

	var outArr = [];
	for(var i = 0; i < arr.length; ++i) {
		outArr.push(encode(arr[i]));
		totalLength += outArr[i].length;
	}

	return Buffer.concat(outArr, totalLength);
}

function decodeNumber(buf, offset, bytes) {
	var negative = bytes < 0;
	bytes = Math.abs(bytes);

	var num = 0;
	var mult = 1;
	var odd;
	for(var i = bytes-1; i >= 0; --i) {
		var b = buf[offset+i];
		if(negative)
			b = -(~b & 0xff);

		if(i == bytes-1)
			odd = b & 0x01;

		num += b * mult;
		mult *= 0x100;
	}

	if(num > maxInt || num < minInt || (num === minInt && odd))
		throw new RangeError('Cannot unpack signed integers larger than 54 bits');

	return num;
}

function decode(buf, pos, nested) {
	if(typeof nested === 'undefined') nested = false;

	var code = buf[pos];
	var value;

	if(code === 0) {
		value = null;
		if(nested) {
			pos += 2;
		} else {
			pos++;
		}
	}
	else if(code === BYTES_CODE || code === STRING_CODE) {
		var nullBytes = findNullBytes(buf, pos+1, true);

		var start = pos+1;
		var end = nullBytes[nullBytes.length-1];

		if(code === STRING_CODE && nullBytes.length === 1) {
			value = buf.toString('utf8', start, end);
		}
		else {
			value = new Buffer(end-start-(nullBytes.length-1));
			var valuePos = 0;

			for(var i=0; i < nullBytes.length && start < end; ++i) {
				buf.copy(value, valuePos, start, nullBytes[i]);
				valuePos += nullBytes[i] - start;
				start = nullBytes[i] + 2;
				if(start <= end) {
					value[valuePos++] = 0;
				}
			}

			if(code === STRING_CODE)
				value = value.toString('utf8');
		}

		pos = end + 1;
	}
	else if(Math.abs(code-INT_ZERO_CODE) <= 7) {
		if(code === INT_ZERO_CODE)
			value = 0;
		else
			value = decodeNumber(buf, pos+1, code-INT_ZERO_CODE);

		pos += Math.abs(INT_ZERO_CODE-code) + 1;
	}
	else if(Math.abs(code-INT_ZERO_CODE) <= 8)
		throw new RangeError('Cannot unpack signed integers larger than 54 bits');
	else if(code === FLOAT_CODE) {
		var valBuf = new Buffer(4);
		buf.copy(valBuf, 0, pos+1, pos+5);
		adjustFloat(valBuf, 0, false);
		value = Float.fromBytes(valBuf);
		pos += 5;
	}
	else if(code === DOUBLE_CODE) {
		var valBuf = new Buffer(8);
		buf.copy(valBuf, 0, pos+1, pos+9);
		adjustFloat(valBuf, 0, false);
		value = Double.fromBytes(valBuf);
		pos += 9;
	}
	else if(code === UUID_CODE) {
		var valBuf = new Buffer(16);
		buf.copy(valBuf, 0, pos+1, pos+17);
		value = new UUID(valBuf);
		pos += 17;
	}
	else if(code === FALSE_CODE) {
		pos++;
		value = false;
	}
	else if(code === TRUE_CODE) {
		pos++;
		value = true;
	}
	else if(code === NESTED_CODE) {
		pos++;
		value = []
		while (buf[pos] != 0 || pos+1 < buf.length && buf[pos+1] === 0xff) {
			var nestedVal = decode(buf, pos, true)
			pos = nestedVal.pos
			value.push(nestedVal.value)
		}
		pos++;
	}
	else
		throw new TypeError('Unknown data type in DB: ' + buf + ' at ' + pos);

	return { pos: pos, value: value };
}

function unpack(key) {
	var res = { pos: 0 };
	var arr = [];

	key = fdbUtil.keyToBuffer(key);

	while(res.pos < key.length) {
		res = decode(key, res.pos);
		arr.push(res.value);
	}

	return arr;
}

function range(arr) {
	var packed = pack(arr);
	return { begin: Buffer.concat([packed, nullByte]), end: Buffer.concat([packed, new Buffer('ff', 'hex')]) };
}

function compare(arr1, arr2) {
	// NOTE: There is built-in comparison function included in 0.11.13 that we might want to switch to.
	var buf1 = pack(arr1);
	var buf2 = pack(arr2);
	var pos = 0;

	while(pos < buf1.length && pos < buf2.length) {
		if(buf1[pos] != buf2[pos]) {
			if(buf1[pos] < buf2[pos]) {
				return -1;
			} else {
				return 1;
			}
		}
		pos += 1;
	}

	// The two arrays begin with a common prefix.
	if(buf1.length < buf2.length) {
		return -1;
	} else if(buf1.length == buf2.length) {
		return 0;
	} else {
		return 1;
	}
}

module.exports = {pack: pack, unpack: unpack, range: range, compare: compare, Float: Float, Double: Double, UUID: UUID};

File: bindings/go/src/fdb/tuple/tuple.go
/*
 * tuple.go
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// FoundationDB Go Tuple Layer

// Package tuple provides a layer for encoding and decoding multi-element tuples
// into keys usable by FoundationDB. The encoded key maintains the same sort
// order as the original tuple: sorted first by the first element, then by the
// second element, etc. This makes the tuple layer ideal for building a variety
// of higher-level data models.
//
// For general guidance on tuple usage, see the Tuple section of Data Modeling
// (https://apple.github.io/foundationdb/data-modeling.html#tuples).
//
// FoundationDB tuples can currently encode byte and unicode strings, integers
// and NULL values. In Go these are represented as []byte, string, int64 and
// nil.
package tuple

import (
	"bytes"
	"encoding/binary"
	"fmt"

	"github.com/apple/foundationdb/bindings/go/src/fdb"
)

// A TupleElement is one of the types that may be encoded in FoundationDB
// tuples. Although the Go compiler cannot enforce this, it is a programming
// error to use an unsupported types as a TupleElement (and will typically
// result in a runtime panic).
//
// The valid types for TupleElement are []byte (or fdb.KeyConvertible), string,
// int64 (or int), float, double, bool, UUID, Tuple, and nil.
type TupleElement interface{}

// Tuple is a slice of objects that can be encoded as FoundationDB tuples. If
// any of the TupleElements are of unsupported types, a runtime panic will occur
// when the Tuple is packed.
//
// Given a Tuple T containing objects only of these types, then T will be
// identical to the Tuple returned by unpacking the byte slice obtained by
// packing T (modulo type normalization to []byte and int64).
type Tuple []TupleElement

// UUID wraps a basic byte array as a UUID. We do not provide any special
// methods for accessing or generating the UUID, but as Go does not provide
// a built-in UUID type, this simple wrapper allows for other libraries
// to write the output of their UUID type as a 16-byte array into
// an instance of this type.
type UUID [16]byte

// Type codes: These prefix the different elements in a packed Tuple
// to indicate what type they are.
const nilCode = 0x00
const bytesCode = 0x01
const stringCode = 0x02
const nestedCode = 0x05
const intZeroCode = 0x14
const posIntEnd = 0x1c
const negIntStart = 0x0c
const floatCode = 0x20
const doubleCode = 0x21
const falseCode = 0x26
const trueCode = 0x27
const uuidCode = 0x30

var sizeLimits = []uint64{
	1<<(0*8) - 1,
	1<<(1*8) - 1,
	1<<(2*8) - 1,
	1<<(3*8) - 1,
	1<<(4*8) - 1,
	1<<(5*8) - 1,
	1<<(6*8) - 1,
	1<<(7*8) - 1,
	1<<(8*8) - 1,
}

func adjustFloatBytes(b []byte, encode bool) {
	if (encode && b[0]&0x80 != 0x00) || (!encode && b[0]&0x80 == 0x00) {
		// Negative numbers: flip all of the bytes.
		for i := 0; i < len(b); i++ {
			b[i] = b[i] ^ 0xff
		}
	} else {
		// Positive number: flip just the sign bit.
		b[0] = b[0] ^ 0x80
	}
}

func encodeBytes(buf *bytes.Buffer, code byte, b []byte) {
	buf.WriteByte(code)
	buf.Write(bytes.Replace(b, []byte{0x00}, []byte{0x00, 0xFF}, -1))
	buf.WriteByte(0x00)
}

func bisectLeft(u uint64) int {
	var n int
	for sizeLimits[n] < u {
		n++
	}
	return n
}

func encodeInt(buf *bytes.Buffer, i int64) {
	if i == 0 {
		buf.WriteByte(0x14)
		return
	}

	var n int
	var ibuf bytes.Buffer

	switch {
	case i > 0:
		n = bisectLeft(uint64(i))
		buf.WriteByte(byte(intZeroCode + n))
		binary.Write(&ibuf, binary.BigEndian, i)
	case i < 0:
		n = bisectLeft(uint64(-i))
		buf.WriteByte(byte(0x14 - n))
		binary.Write(&ibuf, binary.BigEndian, int64(sizeLimits[n])+i)
	}

	buf.Write(ibuf.Bytes()[8-n:])
}

func encodeFloat(buf *bytes.Buffer, f float32) {
	var ibuf bytes.Buffer
	binary.Write(&ibuf, binary.BigEndian, f)
	buf.WriteByte(floatCode)
	out := ibuf.Bytes()
	adjustFloatBytes(out, true)
	buf.Write(out)
}

func encodeDouble(buf *bytes.Buffer, d float64) {
	var ibuf bytes.Buffer
	binary.Write(&ibuf, binary.BigEndian, d)
	buf.WriteByte(doubleCode)
	out := ibuf.Bytes()
	adjustFloatBytes(out, true)
	buf.Write(out)
}

func encodeUUID(buf *bytes.Buffer, u UUID) {
	buf.WriteByte(uuidCode)
	buf.Write(u[:])
}

func encodeTuple(buf *bytes.Buffer, t Tuple, nested bool) {
	if nested {
		buf.WriteByte(nestedCode)
	}

	for i, e := range t {
		switch e := e.(type) {
		case Tuple:
			encodeTuple(buf, e, true)
		case nil:
			buf.WriteByte(nilCode)
			if nested {
				buf.WriteByte(0xff)
			}
		case int64:
			encodeInt(buf, e)
		case int:
			encodeInt(buf, int64(e))
		case []byte:
			encodeBytes(buf, bytesCode, e)
		case fdb.KeyConvertible:
			encodeBytes(buf, bytesCode, []byte(e.FDBKey()))
		case string:
			encodeBytes(buf, stringCode, []byte(e))
		case float32:
			encodeFloat(buf, e)
		case float64:
			encodeDouble(buf, e)
		case bool:
			if e {
				buf.WriteByte(trueCode)
			} else {
				buf.WriteByte(falseCode)
			}
		case UUID:
			encodeUUID(buf, e)
		default:
			panic(fmt.Sprintf("unencodable element at index %d (%v, type %T)", i, t[i], t[i]))
		}
	}

	if nested {
		buf.WriteByte(0x00)
	}
}

// Pack returns a new byte slice encoding the provided tuple. Pack will panic if
// the tuple contains an element of any type other than []byte,
// fdb.KeyConvertible, string, int64, int, float32, float64, bool, tuple.UUID,
// nil, or a Tuple with elements of valid types.
//
// Tuple satisfies the fdb.KeyConvertible interface, so it is not necessary to
// call Pack when using a Tuple with a FoundationDB API function that requires a
// key.
func (t Tuple) Pack() []byte {
	buf := new(bytes.Buffer)
	encodeTuple(buf, t, false)
	return buf.Bytes()
}

func findTerminator(b []byte) int {
	bp := b
	var length int

	for {
		idx := bytes.IndexByte(bp, 0x00)
		length += idx
		if idx+1 == len(bp) || bp[idx+1] != 0xFF {
			break
		}
		length += 2
		bp = bp[idx+2:]
	}

	return length
}

func decodeBytes(b []byte) ([]byte, int) {
	idx := findTerminator(b[1:])
	return bytes.Replace(b[1:idx+1], []byte{0x00, 0xFF}, []byte{0x00}, -1), idx + 2
}

func decodeString(b []byte) (string, int) {
	bp, idx := decodeBytes(b)
	return string(bp), idx
}

func decodeInt(b []byte) (int64, int) {
	if b[0] == intZeroCode {
		return 0, 1
	}

	var neg bool

	n := int(b[0]) - intZeroCode
	if n < 0 {
		n = -n
		neg = true
	}

	bp := make([]byte, 8)
	copy(bp[8-n:], b[1:n+1])

	var ret int64

	binary.Read(bytes.NewBuffer(bp), binary.BigEndian, &ret)

	if neg {
		ret -= int64(sizeLimits[n])
	}

	return ret, n + 1
}

func decodeFloat(b []byte) (float32, int) {
	bp := make([]byte, 4)
	copy(bp, b[1:])
	adjustFloatBytes(bp, false)
	var ret float32
	binary.Read(bytes.NewBuffer(bp), binary.BigEndian, &ret)
	return ret, 5
}

func decodeDouble(b []byte) (float64, int) {
	bp := make([]byte, 8)
	copy(bp, b[1:])
	adjustFloatBytes(bp, false)
	var ret float64
	binary.Read(bytes.NewBuffer(bp), binary.BigEndian, &ret)
	return ret, 9
}

func decodeUUID(b []byte) (UUID, int) {
	var u UUID
	copy(u[:], b[1:])
	return u, 17
}

func decodeTuple(b []byte, nested bool) (Tuple, int, error) {
	var t Tuple

	var i int

	for i < len(b) {
		var el interface{}
		var off int

		switch {
		case b[i] == nilCode:
			if !nested {
				el = nil
				off = 1
			} else if i+1 < len(b) && b[i+1] == 0xff {
				el = nil
				off = 2
			} else {
				return t, i + 1, nil
			}
		case b[i] == bytesCode:
			el, off = decodeBytes(b[i:])
		case b[i] == stringCode:
			el, off = decodeString(b[i:])
		case negIntStart <= b[i] && b[i] <= posIntEnd:
			el, off = decodeInt(b[i:])
		case b[i] == floatCode:
			if i+5 > len(b) {
				return nil, i, fmt.Errorf("insufficient bytes to decode float starting at position %d of byte array for tuple", i)
			}
			el, off = decodeFloat(b[i:])
		case b[i] == doubleCode:
			if i+9 > len(b) {
				return nil, i, fmt.Errorf("insufficient bytes to decode double starting at position %d of byte array for tuple", i)
			}
			el, off = decodeDouble(b[i:])
		case b[i] == trueCode:
			el = true
			off = 1
		case b[i] == falseCode:
			el = false
			off = 1
		case b[i] == uuidCode:
			if i+17 > len(b) {
				return nil, i, fmt.Errorf("insufficient bytes to decode UUID starting at position %d of byte array for tuple", i)
			}
			el, off = decodeUUID(b[i:])
		case b[i] == nestedCode:
			var err error
			el, off, err = decodeTuple(b[i+1:], true)
			if err != nil {
				return nil, i, err
			}
			off++
		default:
			return nil, i, fmt.Errorf("unable to decode tuple element with unknown typecode %02x", b[i])
		}

		t = append(t, el)
		i += off
	}

	return t, i, nil
}

// Unpack returns the tuple encoded by the provided byte slice, or an error if
// the key does not correctly encode a FoundationDB tuple.
func Unpack(b []byte) (Tuple, error) {
	t, _, err := decodeTuple(b, false)
	return t, err
}

// FDBKey returns the packed representation of a Tuple, and allows Tuple to
// satisfy the fdb.KeyConvertible interface. FDBKey will panic in the same
// circumstances as Pack.
func (t Tuple) FDBKey() fdb.Key {
	return t.Pack()
}

// FDBRangeKeys allows Tuple to satisfy the fdb.ExactRange interface. The range
// represents all keys that encode tuples strictly starting with a Tuple (that
// is, all tuples of greater length than the Tuple of which the Tuple is a
// prefix).
func (t Tuple) FDBRangeKeys() (fdb.KeyConvertible, fdb.KeyConvertible) {
	p := t.Pack()
	return fdb.Key(concat(p, 0x00)), fdb.Key(concat(p, 0xFF))
}

// FDBRangeKeySelectors allows Tuple to satisfy the fdb.Range interface. The
// range represents all keys that encode tuples strictly starting with a Tuple
// (that is, all tuples of greater length than the Tuple of which the Tuple is a
// prefix).
func (t Tuple) FDBRangeKeySelectors() (fdb.Selectable, fdb.Selectable) {
	b, e := t.FDBRangeKeys()
	return fdb.FirstGreaterOrEqual(b), fdb.FirstGreaterOrEqual(e)
}

func concat(a []byte, b ...byte) []byte {
	r := make([]byte, len(a)+len(b))
	copy(r, a)
	copy(r[len(a):], b)
	return r
}
